### PR TITLE
Add -skip-oidc-discovery option

### LINKED ISSUE #27: Current OIDC implementation requires provider to implement the /.well-known/openid-configuration endpoint
## Current Behavior

The current implementation expects the OIDC provider to implement the discovery endpoint, however this is optional in the spec and unfortunately some enterprises do not implement this, such as IBM.

## Possible Solution

On my fork of the bitly version, I implemented an additional provider with a fork of the go-oidc package that had an implementation for NewManualProvider, skipping the call to the discovery endpoint.

```
// NewManualProvider creates a provider with manually set configurations
func NewManualProvider(ctx context.Context, issuer, authURL, tokenURL, userInfoURL, jwksURL string) *Provider {
	return &Provider{
		issuer:       issuer,
		authURL:      authURL,
		tokenURL:     tokenURL,
		userInfoURL:  userInfoURL,
		remoteKeySet: newRemoteKeySet(ctx, jwksURL, time.Now),
	}
}
```

We could implement as a new provider or use a flag to let oauth2_proxy know that we are trying to integrate to a provider that does not have the discovery endpoint available.

Alternatively we could handle the error and proceed.

## Steps to Reproduce (for bugs)

Configure the oauth2_proxy against an OIDC provider that does not have the discovery endpoint.

## Context

Some enterprises do not implement or expose the /.well-known/openid-configuration endpoint

## Your Environment

- Version used: bitly's oauth2_proxy with @JoelSpeed's PR's

### PR DESCRIPTION
**Pull request by marratj**
## Description

This PR adds the option to skip OIDC discovery by go-oidc and instead manually configure Authorization, Token and JWKS URI endpoints.

## Motivation and Context

Fixes #27 

Some OIDC providers either do not have a discovery endpoint (non-standard Enterprise solutions) or require additional URL parameters for specific user journey policies to be supplied (Microsoft Azure AD B2C). 

go-oidc calls the fixed `/.well-known/openid-configuration` endpoint of the configured OIDC issuer. When there is no issuer or when there are multiple policies defined via additional parameters, this method cannot be relied upon.

Fortunately, go-oidc provides the possibility to use a custom HTTP client via a client context. With this, we supply go-oidc with a "fake" HTTP client that always returns the manually configured Auth,Token and JWKS endpoints in a JSON response when it calls the discovery endpoint.

## How Has This Been Tested?

I tested this with Microsoft Azure AD B2C as OIDC Provider which uses an additional URL parameter for custom User Journeys. Manually setting the OIDC endpoints to a particular policy returns the proper user journey

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by marratj** [comment]
@tlawrie care to test this PR in your environment?

File: options.go [hunk 1]
```diff
@@ -133,6 +140,32 @@ func parseURL(toParse string, urltype string, msgs []string) (*url.URL, []string
 	return parsed, msgs
 }
 
+// OIDCManualTransport is a structu that is used for manual OIDC endpoints definition
```
Comment (aslafy-z) [inline options.go]:
```suggestion
// OIDCManualTransport is a structure that is used for manual OIDC endpoints definition
```

File: options.go [hunk 2]
```diff
@@ -70,6 +73,9 @@ type Options struct {
 	// potential overrides.
 	Provider          string `flag:"provider" cfg:"provider"`
 	OIDCIssuerURL     string `flag:"oidc-issuer-url" cfg:"oidc_issuer_url"`
+	SkipOidcDiscovery bool   `flag:"skip-oidc-discovery" cfg:"skip_oidc_discovery"`
+	OIDCUserInfoURL   string `flag:"oidc-userinfo-url" cfg:"oidc_userinfo_url"`
```
Comment (JoelSpeed) [inline options.go]: I don't think this is needed? Is the userinfo URL actually used anywhere?

File: options.go [hunk 3]
```diff
@@ -70,6 +73,9 @@ type Options struct {
 	// potential overrides.
 	Provider          string `flag:"provider" cfg:"provider"`
 	OIDCIssuerURL     string `flag:"oidc-issuer-url" cfg:"oidc_issuer_url"`
+	SkipOidcDiscovery bool   `flag:"skip-oidc-discovery" cfg:"skip_oidc_discovery"`
```
Comment (JoelSpeed) [inline options.go]: Can we capitalise OIDC to be consistent
```suggestion
	SkipOIDCDiscovery bool   `flag:"skip-oidc-discovery" cfg:"skip_oidc_discovery"`
```

File: options.go (hunk 2 above)
Comment (marratj) [reply options.go]: agreed, have removed it

File: options.go (hunk 3 above)
Comment (marratj) [reply options.go]: agreed, have changed it

**Comment by marratj** [comment]
rebased to master, using oidc.NewVerifier now (which was actually pretty simple and already documented with an example in go-oidc itself).

@tlawrie care to test the new changes?

**Comment by tlawrie** [comment]
Fantastic. Myself and my colleague @costelmoraru will test

File: options.go [hunk 4]
```diff
@@ -161,16 +164,39 @@ func (o *Options) Validate() error {
...
 
 	if o.OIDCIssuerURL != "" {
-		// Configure discoverable provider data.
-		provider, err := oidc.NewProvider(context.Background(), o.OIDCIssuerURL)
-		if err != nil {
-			return err
+
+		ctx := context.Background()
+
+		// Override discoverable provider data with a custom http.Client that fakes
```
Comment (JoelSpeed) [inline options.go]: This comment is now out of date, could you update it please

Comment (marratj) [reply options.go]: done :)

**Comment by marratj** [comment]
@tlawrie did you have any chance to test yet?
@JoelSpeed when @tlawrie finished his tests, this should be good to merge?

**Comment by costelmoraru** [comment]
Hey @marratj , we tested out the `-skip-oidc-discovery` option on a ICP 2.3.0.1 and ICP 3.1.1 and it works correctly. 

Thank you for your help and contribution.

**Comment by marratj** [comment]
@JoelSpeed @costelmoraru @tlawrie 

I guess we are good to go then?

**Comment by JoelSpeed** [comment]
@marratj Thanks for your work on this! 🎉

**Comment by harindaka** [comment]
@marratj does this mean oauth2_proxy fully supports Azure AD B2C now? Is there any documentation on how I can setup oauth2_proxy to work with B2C? Or should I just refer the documentation for the Azure AD provider at https://pusher.github.io/oauth2_proxy/auth-configuration#azure-auth-provider

**Comment by marratj** [comment]
@harindaka yes, this fully works now, provider to use is the OIDC one, here is a config that we use:

```
provider = "oidc"
oidc_issuer_url = "https://login.microsoftonline.com/{tenant-id}/v2.0/"
scope = "{B2C-app-id} openid offline_access"

skip_oidc_discovery = true
login_url = "https://login.microsoftonline.com/{tenant-id}/oauth2/v2.0/authorize?p=b2c_1a_policyname"
redeem_url = "https://login.microsoftonline.com/{tenant-id}/oauth2/v2.0/token?p=b2c_1a_policyname"
oidc_jwks_url = "https://login.microsoftonline.com/{tenant-id}/discovery/v2.0/keys?p=b2c_1a_policyname"

```

For `{tenant-id}`, use the ID of the tenant (the long UUID one), not the tenant domain (although the latter should work, but we had some issues with that in the past).

**Pull request merged**